sam.c

```diff
@@ -2321,31 +2326,50 @@ int sam_state_destroy(htsFile *fp) {
         if (fd->h) {
             // Notify sam_dispatcher we're closing
             pthread_mutex_lock(&fd->command_m);
-            fd->command = SAM_CLOSE;
+            if (fd->command != SAM_CLOSE_DONE)
+                fd->command = SAM_CLOSE;
             pthread_cond_signal(&fd->command_c);
             ret = -fd->errcode;
-            if (!ret) hts_tpool_wake_dispatch(fd->q); // unstick the reader
+            if (fd->q)
+                hts_tpool_wake_dispatch(fd->q); // unstick the reader
+
+            if (!fp->is_write && fd->q && fd->dispatcher) {
+                for (;;) {
+                    // Avoid deadlocks with dispatcher
+                    if (fd->command == SAM_CLOSE_DONE)
+                        break;
+                    hts_tpool_wake_dispatch(fd->q);
+                    pthread_mutex_unlock(&fd->command_m);
+                    usleep(10000);
+                    pthread_mutex_lock(&fd->command_m);
+                }
+            }
             pthread_mutex_unlock(&fd->command_m);
 
             if (fp->is_write) {
                 // Dispatch the last partial block.
                 sp_bams *gb = fd->curr_bam;
-                if (!ret && gb && gb->nbams > 0)
-                  ret = hts_tpool_dispatch(fd->p, fd->q, sam_format_worker, gb);
+                if (!ret && gb && gb->nbams > 0 && fd->q)
+                    ret = hts_tpool_dispatch(fd->p, fd->q, sam_format_worker, gb);
 
                 // Flush and drain output
-                hts_tpool_process_flush(fd->q);
+                if (fd->q)
+                    hts_tpool_process_flush(fd->q);
                 pthread_mutex_lock(&fd->command_m);
                 if (!ret) ret = -fd->errcode;
                 pthread_mutex_unlock(&fd->command_m);
 
-                while (!ret && !hts_tpool_process_empty(fd->q)) {
+                while (!ret && fd->q && !hts_tpool_process_empty(fd->q)) {
                     usleep(10000);
                     pthread_mutex_lock(&fd->command_m);
-                    if (!ret) ret = -fd->errcode;
+                    ret = -fd->errcode;
+                    // not empty but shutdown implies error
+                    if (hts_tpool_process_is_shutdown(fd->q) && !ret)
```

Comment:
I'm not sure if I understand this bit.  `sam_dispatcher_write()` calls `hts_tpool_process_shutdown()` [on success](https://github.com/samtools/htslib/blob/d99bf404/sam.c#L2792).  That suggests there may be a race between the queue becoming empty, `sam_dispatcher_write()` shutting the queue down and the shutdown check here.  Given that there's also a `usleep()`, I'm surprised that this doesn't always lose.  I think another `!hts_tpool_empty(fd->q)` check  after finding the shutdown state would be needed to ensure that it really was caused by an error.

Reply:
The loop is there to wait for the IO thread (in this case `sam_dispatcher_write()`) to finish.  The flush should have ensured there's no new input and no remaining jobs running, but there may be some output waiting to be written.

Looking at it again, this probably is OK.  The flush will stop new data being added, but the queue is still intact and the writer will patiently wait for more in `hts_tpool_next_result_wait()` until this loop finishes and `hts_tpool_process_shutdown()` [gets called](https://github.com/samtools/htslib/blob/d99bf404/sam.c#L2372).  I suspect the [`hts_tpool_process_shutdown()` call in `sam_dispatcher_write()`](https://github.com/samtools/htslib/blob/d99bf404/sam.c#L2792) isn't actually needed because the queue will already be shut down by the time it gets there.